cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -298,6 +309,70 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 	return nil
 }
 
+// IsCoreDNSConfigMapMigrationRequired checks if a migration of the CoreDNS ConfigMap is required.
+func IsCoreDNSConfigMapMigrationRequired(corefile string) bool {
+	if corefile == "" || migration.Default("", corefile) {
+		return false
+	}
+	return true
+}
+
+func migrateCoreDNSConfigMap(client clientset.Interface, cm *v1.ConfigMap, corefile, currentInstalledCoreDNSVersion string) error {
+	// Since the current Configuration present is the not the default version, try and migrate the Config.
+	updatedCorefile, err := migration.Migrate(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile, false)
+	if err != nil {
+		return errors.Wrap(err, "unable to migrate CoreDNS ConfigMap")
+	}
+
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(&v1.ConfigMap{
```

Comment:
> we can do this:
> - copy the coredns ConfigMap into coredns-backup
> - point the Deployment at the backup
> - wait for update
> - create a new ConfigMap called coredns that has the new config
> point the Deployment at coredns
> - wait for update
> - if the update is not good point the Deployment back at coredns-backup and copy coredns-backup into coredns.
but it requires two rolling updates...
WDYT?

This is somewhat how it is implemented currently except for the last step. 
If creating a separate configmap for backup is not liked like @rosti pointed out, then I can create a Data backup of the Corefile in the same ConfigMap. 

What I suggest: 

- Copy the `Corefile` data as `Corefile-<version>` as backup. 
- Point the Deployment key to the `Corefile-<version>` data and wait for update.
- Migrate the `Corefile` data which will be compatible with the latest CoreDNS release and 
At this point, the ConfigMap will look as follows:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns
  namespace: kube-system
data:
  Corefile:  <Migrated Data>
  Corefile-<version>: <Backup Data>
```
- Now point the Deployment key back to `Corefile` and wait for the update to happen successfully.
- If the update is successful, then we can leave it as-is. 
- If not, we will copy the `Corefile-<version>` to the `Corefile`, notifying the user that migration failed. 
As I understand, this will not trigger the changes immediately but it will ensure that the CoreDNS pods will reload on the changes eventually. 

@chrisohaver @neolit123 @rosti let me know if this makes sense.

@rosti, while this might be a little complicated, the requirement that there be a fail-safe route with at least 1 CoreDNS pod running in the case migration fails.
 [This discussion and requirement](https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-498328908) is what prevented it from this change not happening in k8s 1.15.

Reply:
Makes sense to me.  This lines up with what we had hashed out on the whiteboard last week.  Is it useful to leave the `Corefile-<version>`in place after the migration is over?  It's really there for fall back if the deployment of the migrated Corefile fails.  Certainly if that step fails, then it's not useful and should be removed. It is possible that the step would pass (CoreDNS not crashing), but leave CoreDNS in an otherwise degraded state (e.g. not behaving as expected for some zone) - in which case it might be useful to an admin to reference the previous version of the config.

Reply:
from https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-498328908

> all thats needed for this would be to create the new ConfigMap under the new name, and update the Deployment to use the new ConfigMap. Then Kubernetes rolling update takes over from there.

i was under the impression that we should not create backup `Corefile-back` Data entry, but instead backup a whole ConfigMap and point Deployment at that, but OK.

> Point the Deployment key to the Corefile-<version> data and wait for update.

i guess by patching this part?
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go#L301-L302

>  Is it useful to leave the Corefile-<version>in place after the migration is over? 

1) if we keep a lot of `Corefile-<version>` entries it can inflate the size of the CM indefinitely.
2) if we keep only one version `Corefile-backup` it will allow rollback to an older version, but not further than that.

i'd go for 2, as kubeadm does not support incremental downgrades.

Reply:
Keeping only one previous config makes sense.
We moved to using multiple keys in one map instead of two maps based on feedback (feedback that perhaps we misunderstood).  I think it's just as easy to use separate maps.  I don't have a preference.

Reply:
> i was under the impression that we should not create backup Corefile-back Data entry, but instead backup a whole ConfigMap and point Deployment at that, but OK.

I guess the point of the backup was to ensure that it is done in a way that keeps current pods functioning and rolls out the new config safely.

>i guess by patching this part?
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go#L301-L302

Yes.

Reply:
`Corefile-backup` is good, `Corefile-<version>` will bloat the CM over time.